common/raw_hashtable_metadata_group.h

```diff
@@ -0,0 +1,930 @@
+                                                MetadataGroup rhs) -> bool {
+  return llvm::equal(lhs.bytes, rhs.bytes);
+}
+
+inline auto MetadataGroup::SIMDLoad(const uint8_t* metadata, ssize_t index)
+    -> MetadataGroup {
+  MetadataGroup g;
+#if CARBON_NEON_SIMD_SUPPORT
+  g.byte_vec = vld1_u8(metadata + index);
+#elif CARBON_X86_SIMD_SUPPORT
+  g.byte_vec =
+      _mm_load_si128(reinterpret_cast<const __m128i*>(metadata + index));
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+  return g;
+}
+
+inline auto MetadataGroup::SIMDStore(uint8_t* metadata, ssize_t index) const
+    -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  vst1_u8(metadata + index, byte_vec);
+#elif CARBON_X86_SIMD_SUPPORT
+  _mm_store_si128(reinterpret_cast<__m128i*>(metadata + index), byte_vec);
+#else
+  static_assert(!UseSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(metadata);
+  static_cast<void>(index);
+#endif
+}
+
+inline auto MetadataGroup::SIMDClearDeleted() -> void {
+#if CARBON_NEON_SIMD_SUPPORT
+  byte_ints[0] &= (~LSBs | byte_ints[0] >> 7);
+#elif CARBON_X86_SIMD_SUPPORT
+  byte_vec = _mm_blendv_epi8(_mm_setzero_si128(), byte_vec, byte_vec);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+}
+
+inline auto MetadataGroup::SIMDMatch(uint8_t present_byte) const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto match_byte_vec = vdup_n_u8(present_byte | PresentMask);
+  auto match_byte_cmp_vec = vceq_u8(byte_vec, match_byte_vec);
+  uint64_t mask = vreinterpret_u64_u8(match_byte_cmp_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchRange(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = X86SIMDMatch(present_byte | PresentMask);
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+  static_cast<void>(present_byte);
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchPresent() const -> MatchRange {
+  MatchRange result;
+#if CARBON_NEON_SIMD_SUPPORT
+  // Just extract directly.
+  uint64_t mask = vreinterpret_u64_u8(byte_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchRange(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  // We arrange the byte vector for present bytes so that we can directly
+  // extract it as a mask.
+  result = MatchRange(_mm_movemask_epi8(byte_vec));
+#else
+  static_assert(!UseSIMD && !DebugSIMD, "Unimplemented SIMD operation");
+#endif
+  return result;
+}
+
+inline auto MetadataGroup::SIMDMatchEmpty() const -> MatchIndex {
+  MatchIndex result;
+#if CARBON_NEON_SIMD_SUPPORT
+  auto cmp_vec = vceqz_u8(byte_vec);
+  uint64_t mask = vreinterpret_u64_u8(cmp_vec)[0];
+  // Mask in scalar to fold with testing for zero.
+  result = MatchIndex(mask & MSBs);
+#elif CARBON_X86_SIMD_SUPPORT
+  result = static_cast<MatchIndex>(X86SIMDMatch(Empty));
```

Comment:
Say something about how we aren't zeroing everything but the low bits in this case since we only need to return a `MatchIndex` that captures the first result.

(Also in next function.)

Reply:
We don't do anything special for `MatchIndex` on x86 because there isn't a more efficient approach than what provides the full `MatchRange`. But that seems worth documenting so added comments to both. Does that make sense?